File: configuration-management/examples/go/config/loader_test.go
package config

import (
	"testing"
	"os"
)

func TestLoader_Load(t *testing.T) {
	tests := []struct {
		name    string
		envVars map[string]string
		wantErr bool
	}{
		{
			name: "valid local config",
			envVars: map[string]string{
				"APP_ENV":           "local",
				"DATABASE_HOST":     "localhost",
				"DATABASE_NAME":     "testdb",
				"DATABASE_PASSWORD": "testpass",
				"REDIS_HOST":        "localhost",
			},
			wantErr: false,
		},
		{
			name: "missing required fields",
			envVars: map[string]string{
				"APP_ENV": "local",
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			os.Clearenv()
			for k, v := range tt.envVars {
				os.Setenv(k, v)
			}

			loader := NewLoader()
			cfg, err := loader.Load(nil)

			if (err != nil) != tt.wantErr {
				t.Errorf("Loader.Load() error = %v, wantErr %v", err, tt.wantErr)
				return
			}

			if !tt.wantErr && cfg == nil {
				t.Error("Loader.Load() returned nil config")
			}
		})
	}
}

func TestConfig_Validate(t *testing.T) {
	tests := []struct {
		name    string
		config  *Config
		wantErr bool
	}{
		{
			name: "valid config",
			config: &Config{
				App: AppConfig{
					Name: "myapp",
					Port: 8080,
				},
				Database: DatabaseConfig{
					Host:     "localhost",
					Name:     "testdb",
					Password: "testpass",
					SSLMode:  "require",
				},
				Redis: RedisConfig{
					Host: "localhost",
				},
			},
			wantErr: false,
		},
		{
			name: "missing app name",
			config: &Config{
				App: AppConfig{
					Port: 8080,
				},
				Database: DatabaseConfig{
					Host:     "localhost",
					Name:     "testdb",
					Password: "testpass",
					SSLMode:  "require",
				},
				Redis: RedisConfig{
					Host: "localhost",
				},
			},
			wantErr: true,
		},
		{
			name: "invalid ssl mode",
			config: &Config{
				App: AppConfig{
					Name: "myapp",
					Port: 8080,
				},
				Database: DatabaseConfig{
					Host:     "localhost",
					Name:     "testdb",
					Password: "testpass",
					SSLMode:  "invalid",
				},
				Redis: RedisConfig{
					Host: "localhost",
				},
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := tt.config.Validate()
			if (err != nil) != tt.wantErr {
				t.Errorf("Config.Validate() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestConfig_GetDSN(t *testing.T) {
	cfg := &Config{
		Database: DatabaseConfig{
			Host:     "localhost",
			Port:     5432,
			User:     "testuser",
			Password: "testpass",
			Name:     "testdb",
			SSLMode:  "require",
		},
	}

	dsn := cfg.GetDSN()
	expected := "host=localhost port=5432 user=testuser password=testpass dbname=testdb sslmode=require"
	if dsn != expected {
		t.Errorf("GetDSN() = %v, want %v", dsn, expected)
	}
}

File: configuration-management/examples/nodejs/config/loader.test.js
/**
 * Tests for configuration loader.
 */

const { ConfigLoader, ConfigValidationError, Config, DatabaseConfig, RedisConfig } = require('./loader');

describe('ConfigLoader', () => {
  const originalEnv = process.env;
  
  beforeEach(() => {
    process.env = { ...originalEnv };
  });
  
  afterEach(() => {
    process.env = originalEnv;
  });
  
  describe('load', () => {
    test('should load local configuration', async () => {
      process.env.APP_ENV = 'local';
      process.env.DATABASE_HOST = 'localhost';
      process.env.DATABASE_NAME = 'testdb';
      process.env.DATABASE_PASSWORD = 'testpass';
      process.env.REDIS_HOST = 'localhost';
      
      const loader = new ConfigLoader();
      const config = await loader.load();
      
      expect(config.app.env).toBe('local');
      expect(config.database.host).toBe('localhost');
      expect(config.database.name).toBe('testdb');
      expect(config.redis.host).toBe('localhost');
    });
    
    test('should fail validation for missing required fields', async () => {
      process.env.APP_ENV = 'local';
      delete process.env.DATABASE_HOST;
      
      const loader = new ConfigLoader();
      
      await expect(loader.load()).rejects.toThrow(ConfigValidationError);
    });
    
    test('should override config with env vars', async () => {
      process.env.APP_ENV = 'local';
      process.env.DATABASE_HOST = 'localhost';
      process.env.DATABASE_NAME = 'testdb';
      process.env.DATABASE_PASSWORD = 'testpass';
      process.env.REDIS_HOST = 'localhost';
      process.env.LOG_LEVEL = 'debug';
      process.env.DATABASE_PORT = '5433';
      
      const loader = new ConfigLoader();
      const config = await loader.load();
      
      expect(config.app.logLevel).toBe('debug');
      expect(config.database.port).toBe(5433);
    });
    
    test('should fail for invalid SSL mode', async () => {
      process.env.APP_ENV = 'local';
      process.env.DATABASE_HOST = 'localhost';
      process.env.DATABASE_NAME = 'testdb';
      process.env.DATABASE_PASSWORD = 'testpass';
      process.env.REDIS_HOST = 'localhost';
      process.env.DATABASE_SSL_MODE = 'invalid';
      
      const loader = new ConfigLoader();
      
      await expect(loader.load()).rejects.toThrow('sslMode');
    });
  });
  
  describe('getDSN', () => {
    test('should generate correct DSN', () => {
      const config = new Config({
        database: {
          host: 'localhost',
          port: 5432,
          user: 'testuser',
          password: 'testpass',
          name: 'testdb',
          sslMode: 'require',
        },
      });
      
      const loader = new ConfigLoader();
      const dsn = loader.getDSN(config);
      
      expect(dsn).toContain('host=localhost');
      expect(dsn).toContain('port=5432');
      expect(dsn).toContain('user=testuser');
      expect(dsn).toContain('dbname=testdb');
      expect(dsn).toContain('sslmode=require');
    });
  });
  
  describe('getRedisURL', () => {
    test('should generate correct Redis URL with password', () => {
      const config = new Config({
        redis: {
          host: 'localhost',
          port: 6379,
          password: 'redispass',
          db: 0,
        },
      });
      
      const loader = new ConfigLoader();
      const url = loader.getRedisURL(config);
      
      expect(url).toBe('redis://:redispass@localhost:6379/0');
    });
    
    test('should generate correct Redis URL without password', () => {
      const config = new Config({
        redis: {
          host: 'localhost',
          port: 6379,
          db: 0,
        },
      });
      
      const loader = new ConfigLoader();
      const url = loader.getRedisURL(config);
      
      expect(url).toBe('redis://localhost:6379/0');
    });
  });
});
